Contactfilter, Seperation of groups from contacts...

mightymop opened the pull request:
-add contactfilter for roster
-add collapsibles for rosteritems
-seperate groups from contacts

Deutsch:

-Kontaktfilter hinzugefügt
-Kontaktlisteneinträge zusammenklappbar
-Trennung von Gruppen und Kontakten

mightymop commented:
This PR is related to: 
https://github.com/jsxc/jsxc/issues/957

sualko commented on scss/partials/_roster.scss:
```diff
@@ -301,14 +350,14 @@ li.jsxc-hide-offline {
   overflow: hidden;
   padding: 0;
-  padding-bottom: 15px;
+ /* padding-bottom: 15px;*/
```
Can be removed.

mightymop replied:
removed...

sualko commented on src/ui/Roster.ts:
```diff
@@ -26,19 +26,23 @@ export default class Roster {
       Roster.get();
    }
-
+  
```
One newline too much.

mightymop replied:
removed...

sualko commented on src/ui/Roster.ts:
```diff
@@ -53,6 +57,7 @@ export default class Roster {
       $('.jsxc-js-notice-menu .jsxc-menu__button').text('');
 
       this.contactList = this.element.find('.jsxc-contact-list');
+     this.groupList = this.element.find('.jsxc-group-list');
```
We should run `yarn fix-typescript-format` to fix the code style.

mightymop replied:
inserted space..

sualko commented on src/ui/Roster.ts:
```diff
@@ -79,6 +84,78 @@ export default class Roster {
+      let coll = $(".jsxc-collapsible");
+      let i;
+
+      for (i = 0; i < coll.length; i++)
```
`let` is block scope, so we can use it directly inside the loop declaration.

mightymop replied:
fixed... now using for(let i;...

sualko commented on src/ui/Roster.ts:
```diff
@@ -79,6 +84,78 @@ export default class Roster {
+   
+   private initCollapisible()
+   {
+      let coll = $(".jsxc-collapsible");
```
To be consistent with the rest of the code and improve readability, we should rename this to `collapsibleElement`.

mightymop replied:
renamed coll to collapsibleElement

sualko commented on src/ui/Roster.ts:
```diff
@@ -79,6 +84,78 @@ export default class Roster {
+         {
+            this.classList.toggle("jsxc-active");
+            let content = $(this).next();
+            if ($($(content)[0]).css('display')!=='none'){
```
To get the first element, we can use `content.first()`.

mightymop replied:
ok iam using content.first() now

sualko commented on src/ui/Roster.ts:
```diff
@@ -79,6 +84,78 @@ export default class Roster {
+            this.classList.toggle("jsxc-active");
+            let content = $(this).next();
+            if ($($(content)[0]).css('display')!=='none'){
+                $($(content)[0]).css('display','none');
```
Can probably easier solved with css.

mightymop replied:
how?

sualko commented on src/ui/Roster.ts:
```diff
@@ -79,6 +84,78 @@ export default class Roster {
+   {
+      $("#filterlist").off("keyup").on("keyup",function(e)
+      {
+         $.each( $(".jsxc-contact-list").children("li"), function( key, value )
```
`$(".jsxc-contact-list").children("li").each()` should be easier to read.

mightymop replied:
changed call to $(".jsxc-contact-list").children("li").each() and $(".jsxc-group-list").children("li").each()

sualko commented on src/ui/Roster.ts:
```diff
@@ -79,6 +84,78 @@ export default class Roster {
+      {
+         $.each( $(".jsxc-contact-list").children("li"), function( key, value )
+         {
+             $(this).css({"display":""});
```
Inconsistent with the above code.

mightymop replied:
?

sualko commented on src/ui/Roster.ts:
```diff
@@ -79,6 +84,78 @@ export default class Roster {
+         if ($(this).val())
+            val = new String($(this).val());
+         
+         if (val!=null&&val.length>0)
```
We should use type comparison `!==`. Also this check should not be required, because the val is evaluated before.

mightymop replied:
now using $(this).val().toString()...

sualko commented on src/ui/Roster.ts:
```diff
@@ -127,6 +220,21 @@ export default class Roster {
       rosterItemElement.remove();
+
+      let contactType = contact.getType();
+      let list = contactType=="chat"?this.contactList:this.groupList;
```
Code style should be enforced.

mightymop replied:
now using let list = (contactType==='chat'?this.contactList:this.groupList);

sualko commented on scss/partials/_roster.scss:
```diff
@@ -1,3 +1,52 @@
+.jsxc-collapsible {
+  background-color: #777;
```
With the latest release I introduced css variables to allow to easily adapt the color scheme. Therefore those vars should be used here as well.

mightymop replied:
using var .. now

sualko reviewed (changes requested):
Wow, awesome job. It really looks great.

![2020-12-27 13 33 07 localhost a6daa820b7a1](https://user-images.githubusercontent.com/2974196/103170913-48bb1f00-4848-11eb-9d76-bf048968c1f5.png)

There are only some code style enhancements I would like to apply before merging this pr. Usually travis is reporting most of them, but for some reason travis is quite for all pull requests. Do you want to fix those issues, or is it ok, if I change them? We also need to change the commit message, because we use https://www.conventionalcommits.org/ (they should also be enforced when you are commiting changes, but for some reason this is also not working).

It would also be nice if you could write me an email to klaus@jsxc.org, so we can discuss the way forward.

sualko commented on src/ui/Roster.ts:
```diff
@@ -80,6 +84,78 @@ export default class Roster {
+         //Re init all items to visible, in case of key press is backspace...
+         $('.jsxc-contact-list').children('li').each(function( key, value )
+         {
+             $(this).css({'display':''});
```
That would be the same as `$('.jsxc-contact-list > li').css('display', '');`, which has better performance and is more readable. But anyway, I think we should use classes for this.

mightymop replied:
i've changed it

sualko commented on src/ui/Roster.ts:
```diff
@@ -80,6 +84,78 @@ export default class Roster {
+
+   private addContactFilter()
+   {
+      $('#filterlist').off('keyup').on('keyup',function(e)
```
JSXC is used in many environments, so we should prefix this with `jsxc-`.

mightymop replied:
added jsxc- prefix

sualko commented on src/ui/Roster.ts:
```diff
@@ -270,14 +377,26 @@ export default class Roster {
       let contactName = contact.getName();
 
-      let lastMessageDate = contact.getLastMessageDate();
+      if (!$($(list).parent()[0]).prev().hasClass('jsxc-active'))
```
This is really hard to read.

mightymop replied:
now using this:

    let parentElement = $(list).parent();
      let previousParent = $(parentElement).prev();

      if (!previousParent.hasClass('jsxc-active'))
      {
         parentElement.css('max-height','');
         parentElement.addClass('jsxc-roster-item-filtered');
      }
      else
          if ($($(list).parent()[0]).prev().hasClass('jsxc-active'))
          {
              parentElement.removeClass('jsxc-roster-item-filtered');
              parentElement.css('max-height','100%');
          }

sualko commented on template/roster.hbs:
```diff
@@ -1,33 +1,43 @@
-            <li data-presence="dnd" class="jsxc-dnd">{{t "dnd"}}</li>
-            <li data-presence="offline" class="jsxc-offline">{{t "Offline"}}</li>
-      {{/menu}}
+   <div class="jsxc-real-bottom">
```
Why was the "real" bottom introduced?

mightymop replied:
Because with the changes the Preferencemenu will not stay fixed at the bottom...

sualko reviewed (changes requested):
Looks much better now. Nonetheless I think the css part can still be improved, but I have to take a deeper look.

sualko commented:
I refactored the code a bit and added a search icon to the top. If you click the filter input, you should also see a clear icon on the right. Please test and verify that everything is working as expected.

![image](https://user-images.githubusercontent.com/2974196/103416193-e3d43180-4b85-11eb-8ead-7eba9da78495.png)

sualko commented:
I added you as contributor to this repo, so you can directly push your features as new branch to this repo. This should make things a lot more smooth.

mightymop commented:
> 
> 
> I refactored the code a bit and added a search icon to the top. If you click the filter input, you should also see a clear icon on the right. Please test and verify that everything is working as expected.
> 
> ![image](https://user-images.githubusercontent.com/2974196/103416193-e3d43180-4b85-11eb-8ead-7eba9da78495.png)

looks good to me and runs fine...
the only thing is when i click into the filter the input will resize in its height... Is it wanted like that or not? 
This happens in chrome and firefox.

sualko commented:
Thanks for testing.

> Is it wanted like that or not?

That's intended :smile:
